### PR TITLE
[Gradient Compression] Refactor CommHookInterface and PythonCommHook.

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46566 [Gradient Compression] Add CppCommHook subclass for supporting the C++ API of communication hook.
* **#46512 [Gradient Compression] Refactor CommHookInterface and PythonCommHook.**

1. Merge 1-line PythonCommHook constructor into the header for simplicity.
2. Move the implementation of PythonCommHook destructor from the header file to cpp file.
3. Rename processFuture method as parseHookResult for readability.
4. Simplify some comments.

Original PR issue: C++ DDP Communication Hook https://github.com/pytorch/pytorch/issues/46348

Differential Revision: [D24374282](https://our.internmc.facebook.com/intern/diff/D24374282/)

### CODE REVIEW

File: torch/csrc/distributed/c10d/comm.h [hunk 1]
```diff
@@ -35,45 +35,33 @@ class GradBucket {
...
   virtual c10::intrusive_ptr<torch::jit::Future> runHook(
       const GradBucket& bucket) = 0;
 
-  // Once the grad bucket of Future is ready, c10d reducer will call this
-  // function to get the resulting tensors of the grad bucket. Then c10d
-  // reducer will use these tensors and copy grads to the grads of individual
-  // parameters.
-  virtual std::vector<at::Tensor> processFuture(c10::IValue future_value) = 0;
+  // Returns the resulting tensors once the communication hook result is ready.
+  std::vector<at::Tensor> parseFromHookResult(const c10::IValue& result);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/comm.h:51]: nit: `parseResult` or `parseHookResult` might be sufficient here.

File: torch/csrc/distributed/c10d/comm.h [hunk 2]
```diff
@@ -35,45 +35,33 @@ class GradBucket {
...
   virtual ~CommHookInterface() {}
 
-  // runHook takes a GradBucket type bucket and passes the tensors of
-  // this grad bucket to hook's callback. This function is called once
-  // the bucket is ready. The hook can perform whatever processing is
-  // needed and return a Future that will hold the new value of the grad
-  // bucket's tensors once ready.
+  // Runs the registered communication hook to communicate gradients
+  // asynchronously, Returns a future that holds the communication results.
   virtual c10::intrusive_ptr<torch::jit::Future> runHook(
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/comm.h:47]: Should we preserve the old comments as well? They provide useful information about the hook's contract?

File: torch/csrc/distributed/c10d/comm.h [hunk 3]
```diff
@@ -35,45 +35,33 @@ class GradBucket {
...
       const GradBucket& bucket) = 0;
 
-  // Once the grad bucket of Future is ready, c10d reducer will call this
-  // function to get the resulting tensors of the grad bucket. Then c10d
-  // reducer will use these tensors and copy grads to the grads of individual
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/comm.h:55]: We should preserve these old comments as well indicating how the hook is used by reducer.

File: torch/csrc/distributed/c10d/comm.h [hunk 4]
```diff
@@ -35,45 +35,33 @@ class GradBucket {
...
-// PythonCommHook enables registering a python hook to c10d reducer and is a
-// sub class of CommHookInterface.
 class TORCH_PYTHON_API PythonCommHook : public CommHookInterface {
  public:
-  // The constructor takes a state and a callable hook. Inputs are Python
-  // objects. The state is passed to the hook in runHook function can be used to
-  // maintain and update any state information that users would like to maintain
-  // as part of the training process. The hook can perform whatever processing
-  // user specified and return a Future indicating completion of any async work.
-  PythonCommHook(py::object state, py::object hook);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/comm.h:69]: Preserve old comments.

File: torch/csrc/distributed/c10d/comm.cpp [hunk 5]
```diff
@@ -85,8 +85,23 @@ void broadcast_coalesced(
...
+    py::gil_scoped_acquire ag;
+    py::object obj = torch::jit::toPyObject(result);
+    auto value = torch::jit::toIValue(
+        obj, c10::ListType::create(c10::TensorType::get()));
+
+    return value.toTensorVector();
+  }
+
+  return result.toTensorVector();
+}
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/comm.cpp:104]: This should be a pure virtual method in CommHookInterface such that both PythonCommHook and CppCommHook have their own implementations.

A subclass of CommHookInterface can override runHook to return an IValue of any type (ex: List/Tuple etc.). It's probably best for the subclass to handle the return value and not in the base class itself.

File: torch/csrc/distributed/c10d/comm.h (hunk 1 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/comm.h:51]: Used "parseHookResult". Usually I prefer the naming style of either parseToXXX or parseFromXXX, but I guess parseXXX here should be clear.

File: torch/csrc/distributed/c10d/comm.h (hunk 2 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/comm.h:47]: Feel the old comments are a bit wordy, so rephrased it from 5 lines to 3 lines, which should not really lose any information.

File: torch/csrc/distributed/c10d/comm.cpp [hunk 6]
```diff
@@ -85,8 +85,16 @@ void broadcast_coalesced(
...
+PythonCommHook::~PythonCommHook() {
+  py::gil_scoped_acquire ag;
+  state_.dec_ref();
+  hook_.dec_ref();
+  // Explicitly set state_ and hook_ to nullptr to prevent py::object's dtor
+  // to decref on the PyObject again.
+  // See Note [Destructing py::object] in python_ivalue.h
+  state_.ptr() = nullptr;
+  hook_.ptr() = nullptr;
+}
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/comm.cpp:97]: Thanks for fixing this, did you see some failures that were caused since this was missing?

**Review by pritamdamania87 — COMMENTED** [review]
Code changes look good, although it seems like the windows tests are failing with an error like:

```
comm.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: virtual __cdecl c10d::CommHookInterface::~CommHookInterface(void)" (__imp_??1CommHookInterface@c10d@@UEAA@XZ) referenced in function "public: virtual __cdecl c10d::PythonCommHook::~PythonCommHook(void)" (??1PythonCommHook@c10d@@UEAA@XZ)
init.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl c10d::CommHookInterface::CommHookInterface(void)" (__imp_??0CommHookInterface@c10d@@QEAA@XZ) referenced in function "class std::unique_ptr<class c10d::PythonCommHook,struct std::default_delete<class c10d::PythonCommHook> > __cdecl std::make_unique<class c10d::PythonCommHook,class pybind11::object,class pybind11::object,0>(class pybind11::object &&,class pybind11::object &&)" (??$make_unique@VPythonCommHook@c10d@@Vobject@pybind11@@V34@$0A@@std@@YA?AV?$unique_ptr@VPythonCommHook@c10d@@U?$default_delete@VPythonCommHook@c10d@@@std@@@0@$$QEAVobject@pybind11@@0@Z)
bin\torch_python.dll : fatal error LNK1120: 2 unresolved externals
```

File: torch/csrc/distributed/c10d/comm.cpp (hunk 6 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/comm.cpp:97]: Thanks for catching this!

Realized that there is no easy way to fix this, so I reverted this change. Realized that the definition of both constructor and destructor of a TORCH_API class should only be used in libtorch_core target in build_variables.bzl, but this cpp file belongs to libtorch_python_distributed_core_sources, which is an extension of libtorch_python_sources.

See this picture:
image

On the other hand, not sure if there can be a workaround by somehow extending the file WindowsTorchApiMacro.h (e.g., adding a new specifier like TORCH_CPP_API) @mrshenli  is this possible?

Comment (SciPioneer) [reply torch/csrc/distributed/c10d/comm.cpp]: It seems that reverting this change still couldn't fix the compiler error on Windows platform.

**Review by sab-regime — APPROVED** [review]

**Review by pritamdamania87 — APPROVED** [review]

File: torch/csrc/distributed/c10d/comm.cpp (hunk 6 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/comm.cpp:97]: Replaced TORCH_API by TORCH_PYTHON_API for CommHookInterface.
